File: backend/pkg/gameplugin/sdk/game-plugin/src/binary.ts
import { Client, MessageOption } from './client';
import { channelPublish } from './channel';
import { Envelope } from './protocol';

export const METHOD_BINARY_CREATE = 'binary.create';
export const METHOD_BINARY_WRITE = 'binary.write';
export const METHOD_BINARY_SEAL = 'binary.seal';
export const METHOD_BINARY_READ = 'binary.read';
export const METHOD_BINARY_STAT = 'binary.stat';
export const METHOD_BINARY_RELEASE = 'binary.release';
export const METHOD_BINARY_ABORT = 'binary.abort';

export const DEFAULT_BINARY_CHUNK_SIZE = 512 * 1024;
export const MAX_BINARY_CHUNK_SIZE = 2 * 1024 * 1024;

export type BinaryStorageKind = 'file';
export type BinaryLifetime = 'message' | 'runtime';

export interface BinaryChecksum {
  algorithm: 'sha256' | string;
  value: string;
}

export interface BinaryReference {
  id: string;
  kind: BinaryStorageKind;
  size: number;
  mediaType?: string;
  checksum?: BinaryChecksum;
  lifetime: BinaryLifetime;
  metadata?: Record<string, unknown>;
}

export interface BinaryCreateInput {
  channelId: string;
  expectedSize: number;
  mediaType?: string;
  lifetime?: BinaryLifetime;
  metadata?: Record<string, unknown>;
}

export interface BinaryCreateResult {
  id: string;
  kind: BinaryStorageKind;
  chunkSize: number;
  maxChunkSize: number;
  maxObjectSize: number;
}

export interface BinaryWriteResult {
  id: string;
  written: number;
  nextOffset: number;
}

export interface BinaryReadResult {
  id: string;
  offset: number;
  nextOffset: number;
  data: string;
  eof: boolean;
  size: number;
}

export interface BinaryStatResult {
  id: string;
  kind: BinaryStorageKind;
  size: number;
  mediaType?: string;
  lifetime: BinaryLifetime;
  checksum?: BinaryChecksum;
  metadata?: Record<string, unknown>;
  state: string;
}

function isBinaryObjectId(id: unknown): id is string {
  return typeof id === 'string' && id.length >= 5 && id.length <= 512 && /^bin_[0-9a-f-]+$/.test(id);
}

function requireBinaryReference(ref: BinaryReference): void {
  if (!ref || !isBinaryObjectId(ref.id)) {
    throw new Error('invalid binary reference id');
  }
  if (ref.kind !== 'file') {
    throw new Error(`unsupported binary storage kind '${String(ref.kind)}'`);
  }
  if (!Number.isSafeInteger(ref.size) || ref.size < 0) {
    throw new Error('invalid binary reference size');
  }
  if (ref.lifetime !== 'message' && ref.lifetime !== 'runtime') {
    throw new Error(`invalid binary lifetime '${String(ref.lifetime)}'`);
  }
}

function payloadOf<T>(envelope: Envelope, method: string): T {
  if (envelope.payload === undefined || envelope.payload === null) {
    throw new Error(`${method} returned an empty payload`);
  }
  return envelope.payload as T;
}

export async function binaryCreate(
  client: Client,
  input: BinaryCreateInput,
  opts: MessageOption[] = []
): Promise<BinaryCreateResult> {
  if (!input.channelId || input.channelId.trim() === '') {
    throw new Error('binary.create requires channelId');
  }
  const expectedSize = input.expectedSize;
  if (!Number.isSafeInteger(expectedSize) || expectedSize < 0) {
    throw new Error('binary.create expectedSize must be a non-negative safe integer');
  }
  const envelope = await client.sendReservedRequest(METHOD_BINARY_CREATE, {
    ...input,
    expectedSize,
    lifetime: input.lifetime ?? 'message',
  }, ...opts);
  return payloadOf<BinaryCreateResult>(envelope, METHOD_BINARY_CREATE);
}

export async function binaryWrite(
  client: Client,
  id: string,
  offset: number,
  data: Uint8Array,
  opts: MessageOption[] = []
): Promise<BinaryWriteResult> {
  if (!isBinaryObjectId(id)) throw new Error('binary.write requires a valid binary object id');
  if (!Number.isSafeInteger(offset) || offset < 0) throw new Error('binary.write offset must be a non-negative safe integer');
  if (!(data instanceof Uint8Array) || data.byteLength === 0 || data.byteLength > MAX_BINARY_CHUNK_SIZE) {
    throw new Error(`binary.write chunk must contain 1..${MAX_BINARY_CHUNK_SIZE} bytes`);
  }
  const envelope = client.supportsBinaryFrames()
    ? await client.sendReservedBinaryRequest(METHOD_BINARY_WRITE, id, offset, data, ...opts)
    : await client.sendReservedRequest(METHOD_BINARY_WRITE, {
        id,
        offset,
        data: Buffer.from(data).toString('base64'),
      }, ...opts);
  return payloadOf<BinaryWriteResult>(envelope, METHOD_BINARY_WRITE);
}

export async function binarySeal(
  client: Client,
  id: string,
  opts: MessageOption[] = []
): Promise<BinaryReference> {
  const envelope = await client.sendReservedRequest(METHOD_BINARY_SEAL, { id }, ...opts);
  const result = payloadOf<{ reference: BinaryReference }>(envelope, METHOD_BINARY_SEAL);
  requireBinaryReference(result.reference);
  return result.reference;
}

export async function binaryAbort(client: Client, id: string, opts: MessageOption[] = []): Promise<void> {
  await client.sendReservedRequest(METHOD_BINARY_ABORT, { id }, ...opts);
}

export async function binaryRelease(client: Client, id: string, opts: MessageOption[] = []): Promise<void> {
  await client.sendReservedRequest(METHOD_BINARY_RELEASE, { id }, ...opts);
}

export async function binaryStat(client: Client, id: string, opts: MessageOption[] = []): Promise<BinaryStatResult> {
  const envelope = await client.sendReservedRequest(METHOD_BINARY_STAT, { id }, ...opts);
  return payloadOf<BinaryStatResult>(envelope, METHOD_BINARY_STAT);
}

export async function binaryRead(
  client: Client,
  ref: BinaryReference,
  offset: number,
  maxBytes: number = DEFAULT_BINARY_CHUNK_SIZE,
  opts: MessageOption[] = []
): Promise<{ data: Uint8Array; result: BinaryReadResult }> {
  requireBinaryReference(ref);
  if (!Number.isSafeInteger(offset) || offset < 0) throw new Error('binary.read offset must be a non-negative safe integer');
  if (!Number.isSafeInteger(maxBytes) || maxBytes <= 0 || maxBytes > MAX_BINARY_CHUNK_SIZE) {
    throw new Error(`binary.read maxBytes must be in 1..${MAX_BINARY_CHUNK_SIZE}`);
  }
  const envelope = await client.sendReservedRequest(METHOD_BINARY_READ, {
    reference: ref,
    offset,
    maxBytes,
  }, ...opts);
  const result = payloadOf<BinaryReadResult>(envelope, METHOD_BINARY_READ);
  return { data: Uint8Array.from(Buffer.from(result.data, 'base64')), result };
}

export async function binaryUpload(
  client: Client,
  input: BinaryCreateInput,
  data: Uint8Array,
  opts: MessageOption[] = []
): Promise<BinaryReference> {
  const created = await binaryCreate(client, { ...input, expectedSize: data.byteLength }, opts);
  let sealed = false;
  try {
    let chunkSize = created.chunkSize;
    if (!Number.isSafeInteger(chunkSize) || chunkSize <= 0 || chunkSize > created.maxChunkSize) {
      chunkSize = Math.min(DEFAULT_BINARY_CHUNK_SIZE, created.maxChunkSize || MAX_BINARY_CHUNK_SIZE);
    }
    chunkSize = Math.min(chunkSize, MAX_BINARY_CHUNK_SIZE);
    let offset = 0;
    while (offset < data.byteLength) {
      const end = Math.min(offset + chunkSize, data.byteLength);
      const result = await binaryWrite(client, created.id, offset, data.subarray(offset, end), opts);
      if (result.nextOffset <= offset) throw new Error(`binary.write made no progress at offset ${offset}`);
      offset = result.nextOffset;
    }
    const ref = await binarySeal(client, created.id, opts);
    sealed = true;
    return ref;
  } finally {
    if (!sealed) {
      try { await binaryAbort(client, created.id, opts); } catch { /* best-effort cleanup */ }
    }
  }
}

export async function binaryReadAll(
  client: Client,
  ref: BinaryReference,
  opts: MessageOption[] = []
): Promise<Uint8Array> {
  requireBinaryReference(ref);
  const chunks: Uint8Array[] = [];
  let total = 0;
  let offset = 0;
  while (offset < ref.size) {
    const { data, result } = await binaryRead(client, ref, offset, DEFAULT_BINARY_CHUNK_SIZE, opts);
    chunks.push(data);
    total += data.byteLength;
    if (result.nextOffset <= offset && !result.eof) throw new Error(`binary.read made no progress at offset ${offset}`);
    offset = result.nextOffset;
    if (result.eof) break;
  }
  if (total !== ref.size) throw new Error(`binary.read size mismatch: got ${total}, want ${ref.size}`);
  const out = new Uint8Array(total);
  let cursor = 0;
  for (const chunk of chunks) {
    out.set(chunk, cursor);
    cursor += chunk.byteLength;
  }
  return out;
}

export async function channelPublishBinary(
  client: Client,
  channelId: string,
  ref: BinaryReference,
  metadata?: Record<string, unknown>,
  opts: MessageOption[] = []
): Promise<Envelope> {
  requireBinaryReference(ref);
  return channelPublish(client, { channelId, payload: ref, metadata }, opts);
}

File: .baseline-check/sdk/game-plugin-go/binary.go
package sdk

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"

	"github.com/u-ai/game-plugin-sdk-go/protocol"
)

const (
	MethodBinaryCreate  = "binary.create"
	MethodBinaryWrite   = "binary.write"
	MethodBinarySeal    = "binary.seal"
	MethodBinaryRead    = "binary.read"
	MethodBinaryStat    = "binary.stat"
	MethodBinaryRelease = "binary.release"
	MethodBinaryAbort   = "binary.abort"

	DefaultBinaryChunkSize = 512 * 1024
	MaxBinaryChunkSize     = 2 * 1024 * 1024
)

type BinaryReference = protocol.BinaryReference
type BinaryStorageKind = protocol.BinaryStorageKind
type BinaryLifetime = protocol.BinaryLifetime
type BinaryChecksum = protocol.BinaryChecksum

const (
	BinaryStorageFile     = protocol.BinaryStorageFile
	BinaryLifetimeMessage = protocol.BinaryLifetimeMessage
	BinaryLifetimeRuntime = protocol.BinaryLifetimeRuntime
)

type BinaryCreateInput struct {
	ChannelID    string                     `json:"channelId"`
	ExpectedSize int64                      `json:"expectedSize"`
	MediaType    string                     `json:"mediaType,omitempty"`
	Lifetime     BinaryLifetime             `json:"lifetime,omitempty"`
	Metadata     map[string]json.RawMessage `json:"metadata,omitempty"`
}

type BinaryCreateResult struct {
	ID            string            `json:"id"`
	Kind          BinaryStorageKind `json:"kind"`
	ChunkSize     int               `json:"chunkSize"`
	MaxChunkSize  int               `json:"maxChunkSize"`
	MaxObjectSize int64             `json:"maxObjectSize"`
}

type BinaryWriteResult struct {
	ID         string `json:"id"`
	Written    int    `json:"written"`
	NextOffset int64  `json:"nextOffset"`
}

type BinarySealResult struct {
	Reference BinaryReference `json:"reference"`
}

type BinaryReadResult struct {
	ID         string `json:"id"`
	Offset     int64  `json:"offset"`
	NextOffset int64  `json:"nextOffset"`
	Data       string `json:"data"`
	EOF        bool   `json:"eof"`
	Size       int64  `json:"size"`
}

type BinaryStatResult struct {
	ID        string                     `json:"id"`
	Kind      BinaryStorageKind          `json:"kind"`
	Size      int64                      `json:"size"`
	MediaType string                     `json:"mediaType,omitempty"`
	Lifetime  BinaryLifetime             `json:"lifetime"`
	Checksum  *BinaryChecksum            `json:"checksum,omitempty"`
	Metadata  map[string]json.RawMessage `json:"metadata,omitempty"`
	State     string                     `json:"state"`
}

func (c *Client) BinaryCreate(ctx context.Context, input BinaryCreateInput, opts ...MessageOption) (BinaryCreateResult, error) {
	if input.Lifetime == "" {
		input.Lifetime = BinaryLifetimeMessage
	}
	var out BinaryCreateResult
	if err := c.binaryRequest(ctx, MethodBinaryCreate, input, &out, opts...); err != nil {
		return BinaryCreateResult{}, err
	}
	return out, nil
}

func (c *Client) BinaryWrite(ctx context.Context, id string, offset int64, data []byte, opts ...MessageOption) (BinaryWriteResult, error) {
	if len(data) == 0 {
		return BinaryWriteResult{}, fmt.Errorf("binary write chunk must not be empty")
	}
	if len(data) > MaxBinaryChunkSize {
		return BinaryWriteResult{}, fmt.Errorf("binary write chunk exceeds %d bytes", MaxBinaryChunkSize)
	}
	if offset < 0 {
		return BinaryWriteResult{}, fmt.Errorf("binary write offset must not be negative")
	}
	if fast, ok := c.transport.(BinaryFrameTransport); ok {
		return c.binaryWriteFrame(ctx, fast, id, offset, data, opts...)
	}
	input := map[string]any{
		"id":     id,
		"offset": offset,
		"data":   base64.StdEncoding.EncodeToString(data),
	}
	var out BinaryWriteResult
	if err := c.binaryRequest(ctx, MethodBinaryWrite, input, &out, opts...); err != nil {
		return BinaryWriteResult{}, err
	}
	return out, nil
}

func (c *Client) binaryWriteFrame(ctx context.Context, transport BinaryFrameTransport, id string, offset int64, data []byte, opts ...MessageOption) (BinaryWriteResult, error) {
	envelope, err := c.NewRequest(MethodBinaryWrite, nil, opts...)
	if err != nil {
		return BinaryWriteResult{}, err
	}
	timeout := c.pendingTimeoutMs
	if envelope.Metadata != nil {
		if raw, ok := envelope.Metadata["__timeout"]; ok {
			var ms int
			if json.Unmarshal(raw, &ms) == nil && ms > 0 {
				timeout = time.Duration(ms) * time.Millisecond
			}
		}
	}
	pending := c.registerPending(envelope.ID, MethodBinaryWrite, timeout)
	defer c.removePending(envelope.ID)
	if err := transport.SendBinaryFrame(ctx, envelope, id, offset, data); err != nil {
		return BinaryWriteResult{}, NewTransportError("send binary frame failed: %v", err)
	}

	select {
	case <-pending.Done:
		pending.mu.Lock()
		state := pending.state
		resp := pending.Response
		pendingErr := pending.Err
		pending.mu.Unlock()
		if state != stateCompleted {
			if pendingErr != nil {
				return BinaryWriteResult{}, pendingErr
			}
			return BinaryWriteResult{}, NewTransportError("binary write request %s ended in state %d", envelope.ID, state)
		}
		if pendingErr != nil {
			return BinaryWriteResult{}, pendingErr
		}
		var out BinaryWriteResult
		if len(resp.Payload) == 0 {
			return BinaryWriteResult{}, NewEncodeError("binary.write returned an empty response")
		}
		if err := json.Unmarshal(resp.Payload, &out); err != nil {
			return BinaryWriteResult{}, NewEncodeError("unmarshal binary.write response: %v", err)
		}
		return out, nil
	case <-ctx.Done():
		return BinaryWriteResult{}, NewTransportError("binary write request %s cancelled: %v", envelope.ID, ctx.Err())
	}
}

func (c *Client) BinarySeal(ctx context.Context, id string, opts ...MessageOption) (BinaryReference, error) {
	var out BinarySealResult
	if err := c.binaryRequest(ctx, MethodBinarySeal, map[string]any{"id": id}, &out, opts...); err != nil {
		return BinaryReference{}, err
	}
	if err := out.Reference.Validate(); err != nil {
		return BinaryReference{}, fmt.Errorf("invalid binary reference returned by host: %w", err)
	}
	return out.Reference, nil
}

func (c *Client) BinaryAbort(ctx context.Context, id string, opts ...MessageOption) error {
	return c.binaryRequest(ctx, MethodBinaryAbort, map[string]any{"id": id}, nil, opts...)
}

func (c *Client) BinaryRelease(ctx context.Context, id string, opts ...MessageOption) error {
	return c.binaryRequest(ctx, MethodBinaryRelease, map[string]any{"id": id}, nil, opts...)
}

func (c *Client) BinaryStat(ctx context.Context, id string, opts ...MessageOption) (BinaryStatResult, error) {
	var out BinaryStatResult
	if err := c.binaryRequest(ctx, MethodBinaryStat, map[string]any{"id": id}, &out, opts...); err != nil {
		return BinaryStatResult{}, err
	}
	return out, nil
}

func (c *Client) BinaryRead(ctx context.Context, ref BinaryReference, offset int64, maxBytes int, opts ...MessageOption) ([]byte, BinaryReadResult, error) {
	if maxBytes <= 0 {
		maxBytes = DefaultBinaryChunkSize
	}
	if maxBytes > MaxBinaryChunkSize {
		return nil, BinaryReadResult{}, fmt.Errorf("binary read chunk exceeds %d bytes", MaxBinaryChunkSize)
	}
	input := map[string]any{"reference": ref, "offset": offset, "maxBytes": maxBytes}
	var out BinaryReadResult
	if err := c.binaryRequest(ctx, MethodBinaryRead, input, &out, opts...); err != nil {
		return nil, BinaryReadResult{}, err
	}
	data, err := base64.StdEncoding.DecodeString(out.Data)
	if err != nil {
		return nil, BinaryReadResult{}, fmt.Errorf("host returned invalid base64 binary chunk: %w", err)
	}
	return data, out, nil
}

// BinaryUpload is the normal high-level upload path. It creates a host-owned
// binary object, uploads bounded chunks, and seals it into a BinaryReference.
// Any failure aborts the partial object before returning.
func (c *Client) BinaryUpload(ctx context.Context, input BinaryCreateInput, data []byte, opts ...MessageOption) (BinaryReference, error) {
	input.ExpectedSize = int64(len(data))
	created, err := c.BinaryCreate(ctx, input, opts...)
	if err != nil {
		return BinaryReference{}, err
	}
	committed := false
	defer func() {
		if !committed {
			_ = c.BinaryAbort(context.Background(), created.ID)
		}
	}()

	chunkSize := created.ChunkSize
	if chunkSize <= 0 || chunkSize > created.MaxChunkSize {
		chunkSize = DefaultBinaryChunkSize
	}
	if created.MaxChunkSize > 0 && chunkSize > created.MaxChunkSize {
		chunkSize = created.MaxChunkSize
	}
	if chunkSize > MaxBinaryChunkSize {
		chunkSize = MaxBinaryChunkSize
	}

	offset := int64(0)
	for len(data) > 0 {
		n := chunkSize
		if len(data) < n {
			n = len(data)
		}
		result, err := c.BinaryWrite(ctx, created.ID, offset, data[:n], opts...)
		if err != nil {
			return BinaryReference{}, err
		}
		offset = result.NextOffset
		data = data[n:]
	}
	ref, err := c.BinarySeal(ctx, created.ID, opts...)
	if err != nil {
		return BinaryReference{}, err
	}
	committed = true
	return ref, nil
}

func (c *Client) BinaryReadAll(ctx context.Context, ref BinaryReference, opts ...MessageOption) ([]byte, error) {
	if err := ref.Validate(); err != nil {
		return nil, err
	}
	if ref.Size < 0 || ref.Size > int64(^uint(0)>>1) {
		return nil, fmt.Errorf("binary object is too large for this process")
	}
	out := make([]byte, 0, int(ref.Size))
	offset := int64(0)
	for offset < ref.Size {
		chunk, result, err := c.BinaryRead(ctx, ref, offset, DefaultBinaryChunkSize, opts...)
		if err != nil {
			return nil, err
		}
		out = append(out, chunk...)
		if result.NextOffset <= offset && !result.EOF {
			return nil, fmt.Errorf("binary read made no progress at offset %d", offset)
		}
		offset = result.NextOffset
		if result.EOF {
			break
		}
	}
	if int64(len(out)) != ref.Size {
		return nil, fmt.Errorf("binary read size mismatch: got %d, want %d", len(out), ref.Size)
	}
	return out, nil
}

func (c *Client) ChannelPublishBinary(ctx context.Context, channelID string, ref BinaryReference, metadata map[string]json.RawMessage, opts ...MessageOption) (protocol.Envelope, error) {
	if err := ref.Validate(); err != nil {
		return protocol.Envelope{}, err
	}
	payload, err := json.Marshal(ref)
	if err != nil {
		return protocol.Envelope{}, err
	}
	return c.ChannelPublish(ctx, ChannelPublishInput{ChannelID: channelID, Payload: payload, Metadata: metadata}, opts...)
}

func (c *Client) binaryRequest(ctx context.Context, method string, input any, output any, opts ...MessageOption) error {
	envelope, err := c.SendReservedRequest(ctx, method, input, opts...)
	if err != nil {
		return err
	}
	if output == nil || len(envelope.Payload) == 0 {
		return nil
	}
	if err := json.Unmarshal(envelope.Payload, output); err != nil {
		return NewEncodeError("unmarshal %s response: %v", method, err)
	}
	return nil
}
